pybandits/simulator.py

```diff
@@ -0,0 +1,592 @@
+
+        def model_post_init(self, __context: Any) -> None:
+            # set random seed for reproducibility
+            random.seed(self.random_seed)
+            np.random.default_rng(self.random_seed)
+            self._initialize_results()
+
+    else:
+        raise ValueError(f"Unsupported pydantic version: {pydantic_version}")
+
+    @abstractmethod
+    def _initialize_results(self):
+        """
+        Initialize the results DataFrame. The results DataFrame is used to store the raw simulation results.
+        """
+        pass
+
+    @property
+    def results(self):
+        return self._results
+
+    def run(self):
+        """
+        Start simulation process. It consists in the following steps:
+            for i=0 to n_updates
+                Consider batch[i] of observation
+                mab selects the best action as the action with the highest reward probability to each sample in
+                    batch[i].
+                Rewards are returned for each recommended action
+                Prior parameters are updated based on recommended actions and returned rewards
+        """
+        for batch_index in range(self.n_updates):
+            predict_kwargs, update_kwargs, metadata = self._get_batch_step_kwargs_and_metadata(batch_index)
+            self._step(batch_index, metadata, predict_kwargs, update_kwargs)
+
+        self._finalize_results()
+
+        # print results
+        if self.verbose:
+            self._print_results()
+
+        if self.visualize:
+            self._visualize_results()
+
+        # store results
+        if self.save:
+            if self.verbose:
+                logger.info(f"Saving results at {self.path}")
+            self._save_results()
+
+    def _step(
+        self,
+        batch_index: int,
+        metadata: Dict[str, List],
+        predict_kwargs: Dict[str, Union[int, np.ndarray]],
+        update_kwargs: Dict[str, np.ndarray],
+    ):
+        """
+        Perform a step of the simulation process. It consists in the following steps:
+            - select actions for batch via mab.predict
+            - draw rewards for the selected actions based on metadata according to probs_reward
+            - write the selected actions for batch #i in the results matrix
+            - update the mab model with the selected actions and the corresponding rewards via mab.update
+
+        Parameters
+        ----------
+        batch_index : int
+            The index of the batch.
+        metadata : Dict[str, List]
+            The metadata for the selected actions.
+        predict_kwargs : Dict[str, Union[int, np.ndarray]]
+            Dictionary containing the keyword arguments for the batch used in mab.predict.
+        update_kwargs : Dict[str, np.ndarray]
+            Dictionary containing the keyword arguments for the batch used in mab.update.
+        """
+        # select actions for batch #index
+        actions = self.mab.predict(**predict_kwargs)[0]
```

Comment:
This is not really related to this PR, but I got confused about why you did:
`mab.predict()[0]` (which is correct) instead of `actions, _ = mab.predict()`.

The reason you did that is that [SmabPredictions](https://github.com/PlaytikaOSS/pybandits/blob/bd54568121b63d2566e56b46c64b2790e4040c83/pybandits/base.py#L39C1-L42C2) and [CmabPredictions](https://github.com/PlaytikaOSS/pybandits/blob/bd54568121b63d2566e56b46c64b2790e4040c83/pybandits/base.py#L40C1-L40C16) are different. Shall we add in `base.py` a description of what are they? E.g.:

```
# SmabPredictions is a tuple containing: the list of the predicted action ID; the list of the sampled reward probability for all possible action IDs
SmabPredictions = NewType(...

# CmabPredictions is a tuple containing: the list of the predicted action ID; the list of the sampled reward probability for all possible action IDs; the list of the sampled reward weighted sum for all possible action IDs
CmabPredictions = NewType(...
```

Reply:
Done

Reply:
I also added comment Simulator line #231